Il2CppInspector.Common/Reflection/TypeInfo.cs

```diff
@@ -769,22 +742,30 @@ public class TypeInfo : MemberInfo
         public TypeInfo(MethodBase declaringMethod, Il2CppGenericParameter param) : this(declaringMethod.DeclaringType, param)
             => DeclaringMethod = declaringMethod;
 
-        // Initialize a type that is a reference to the specified type
-        private TypeInfo(TypeInfo underlyingType) {
-            ElementType = underlyingType;
-            IsByRef = true;
+        // Initialize a type that is an array of the specified type
+        private TypeInfo(TypeInfo elementType, int rank) : base(elementType.Assembly) {
+            ElementType = elementType;
+            IsArray = true;
 
-            // No base type or declaring type for reference types
-            Assembly = ElementType.Assembly;
-            Definition = ElementType.Definition;
-            Index = ElementType.Index;
             Namespace = ElementType.Namespace;
             Name = ElementType.Name;
+            arrayRank = rank;
+        }
 
-            Attributes = ElementType.Attributes;
+        // Initialize a type that is a reference or pointer to the specified type
+        private TypeInfo(TypeInfo underlyingType, bool isPointer) : base(underlyingType.Assembly) {
+            ElementType = underlyingType;
+            if(isPointer) {
+                IsPointer = true;
+            } else {
+                IsByRef = true;
+            }
+
+            Namespace = ElementType.Namespace;
+            Name = ElementType.Name;
         }
 
-        public TypeInfo MakeByRefType() => new TypeInfo(this);
+        public TypeInfo MakeByRefType() => new TypeInfo(this, isPointer: false);
 
```

Comment:
I like the refactoring but something is wrong; this causes #11 to fail once IDA script output is added to the test runner in a later commit. A null type manages to sneak its way into Il2CppModel.Types causing a `NullReferenceException` here:

```cs
        private void writeMethods() {
            writeSectionHeader("Method definitions");
            foreach (var type in model.Types) {
                writeMethods(type.Name, type.DeclaredConstructors);
                writeMethods(type.Name, type.DeclaredMethods);
            }
```
Here is the ida.py output; I added a null check that writes NULL TYPE instead of crashing but you should find out why this happens in case there is any information missing:

```
SetName(0x01692494, r'ValueTuple`2[Boolean,ValueTuple`2[Boolean,ValueTuple`2[Boolean,ValueTuple`2[Boolean,ValueTuple`2[Boolean,UniRx.Async.AsyncUnit]]]]]_System.IValueTupleInternal.ToStringEnd')
idc.set_cmt(0x01692494, r'String System.IValueTupleInternal.ToStringEnd()', 1)
SetName(0x0169249C, r'ValueTuple`2[Boolean,ValueTuple`2[Boolean,ValueTuple`2[Boolean,ValueTuple`2[Boolean,ValueTuple`2[Boolean,UniRx.Async.AsyncUnit]]]]]_System.Runtime.CompilerServices.ITuple.get_Length')
idc.set_cmt(0x0169249C, r'Int32 System.Runtime.CompilerServices.ITuple.get_Length()', 1)
NULL TYPE
SetName(0x0206F204, r'LazyPromise`1[System.ValueTuple`2]_.ctor')
idc.set_cmt(0x0206F204, r'LazyPromise`1[System.ValueTuple`2](Func`1[UniRx.Async.UniTask`1])', 1)
SetName(0x0206F23C, r'LazyPromise`1[System.ValueTuple`2]_Create')
idc.set_cmt(0x0206F23C, r'Void Create()', 1)
```